### PR TITLE
Run tasks on init

### PR DESCRIPTION
**Pull request by rhuttman**

### CODE REVIEW

File: pallets/automation-time/src/lib.rs [hunk 1]
```diff
@@ -251,5 +292,56 @@ pub mod pallet {
...
+		/// TODO (ENG-157):
+		/// - Calculate weights for each task
+		pub fn run_tasks(max_weight: Weight) -> Weight {
+			let mut weight_left: Weight = max_weight - RUN_TASK_OVERHEAD;
+			let time_block = Self::get_current_time_slot();
+
+			if let Some(task_ids) = Self::get_scheduled_tasks(time_block) {
+				let mut consumed_task_index: usize = 0;
+				for task_id in task_ids.iter() {
+					if weight_left < MAX_LOOP_WEIGHT {
```
Comment (irsal) [inline pallets/automation-time/src/lib.rs:313]: good place to throw an err (whenever we figure that out)

File: pallets/automation-time/src/lib.rs [hunk 2]
```diff
@@ -251,5 +292,56 @@ pub mod pallet {
...
+				}
+			}
+
+			max_weight - weight_left
+		}
+
+		/// Fire the notify event with the custom message.
+		/// TODO: Calculate weight (ENG-157).
+		fn run_notify_task(message: Vec<u8>) -> Weight {
+			Self::deposit_event(Event::Notify { message });
```
Comment (irsal) [inline pallets/automation-time/src/lib.rs:343]: this goes tot he treasury right?

File: pallets/automation-time/src/mock.rs [hunk 3]
```diff
@@ -71,14 +72,21 @@ impl pallet_timestamp::Config for Test {
 
 parameter_types! {
 	pub const MaxTasksPerSlot: u32 = 2;
+	pub const MaxBlockWeight: Weight = 600_000;
```
Comment (irsal) [inline pallets/automation-time/src/mock.rs:75]: where'd we get this?

File: pallets/automation-time/src/tests.rs [hunk 4]
```diff
@@ -1,7 +1,5 @@
-// use super::*;
-// use mock::*;
 use crate::{mock::*, Error, Task};
-use frame_support::{assert_noop, assert_ok};
+use frame_support::{assert_noop, assert_ok, traits::OnInitialize};
```
Comment (irsal) [inline pallets/automation-time/src/tests.rs:2]: Would be nice to see a test for the edge case where we overload the task list (ie we're at capacity).

File: pallets/automation-time/src/tests.rs [hunk 5]
```diff
@@ -174,3 +265,11 @@ fn create_task(owner: AccountId, scheduled_time: u64, message: Vec<u8>) -> sp_co
 	let task_ids = AutomationTime::get_scheduled_tasks(scheduled_time).unwrap();
 	task_ids[task_ids.len() - 1]
 }
+
+fn events() -> Vec<Event> {
```
Comment (irsal) [inline pallets/automation-time/src/tests.rs:269]: explain to me please?

**Review by irsal — COMMENTED** [review]
Mostly questions, non-blocking. Trying to understand further.

File: pallets/automation-time/src/lib.rs [hunk 6]
```diff
@@ -35,20 +35,35 @@ mod mock;
...
 
 #[frame_support::pallet]
 pub mod pallet {
 	use super::*;
 
 	type AccountOf<T> = <T as frame_system::Config>::AccountId;
 
+	/// The static weight to run scheduled tasks.
+	/// TODO: Calculate (ENG-157).
+	const RUN_TASK_OVERHEAD: Weight = 30_000;
```
[the remark this answers is not in the document]
Comment (rhuttman) [reply pallets/automation-time/src/lib.rs:54]: 100% made them up for now. I just need values in to test with.

File: pallets/automation-time/src/lib.rs (hunk 1 above)
Comment (rhuttman) [reply pallets/automation-time/src/lib.rs:313]: well we expect this to happen a lot of the time. The overflow case (which I still need to code) would for sure need the error

File: pallets/automation-time/src/lib.rs (hunk 2 above)
Comment (rhuttman) [reply pallets/automation-time/src/lib.rs:343]: I am not sure what you mean. This sends an event. It does not move tokens around or anything

File: pallets/automation-time/src/mock.rs (hunk 3 above)
Comment (rhuttman) [reply pallets/automation-time/src/mock.rs:75]: I just need a value to model test cases after

File: pallets/automation-time/src/tests.rs (hunk 4 above)
Comment (rhuttman) [reply pallets/automation-time/src/tests.rs:2]: I will add that to the overflow tests. Right now the code stops that from happening by WAY under allocating how many tasks it can do per block

File: pallets/automation-time/src/lib.rs (hunk 2 above)
Comment (irsal) [reply pallets/automation-time/src/lib.rs:343]: there's also a depost funds event for transaction fees, i think; i was confusing that w/ this

**Review by irsal — APPROVED** [review]
👍
